02-peer-protocol.md

```diff
@@ -430,6 +446,515 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+  - otherwise:
+    - MUST use the sent and received `input_info` and `output_info`
+      to create the funding transaction, using `max_witness_len`
+      for each `input_info` and `feerate_per_kw_funding` as specified in
+      `open_channel2`.
+    - MUST send `commitment_signed`.
+
+
+##### Rationale
+
+Each node must have a complete set of the transaction inputs and outputs,
+to derive the funding transaction and subsequent commitment signatures.
+
+`funding_add_complete` concludes a node's transmission of its contributions
+to the funding transaction, which serve as an early check that all
+`funding_add_input` and `funding_add_output` messages have been received.
+
+Upon successful exchange of `funding_add_complete` messages, both nodes
+should build the funding transaction and update their channel id
+to be transmitted in `commitment_signed`.
+
+
+### The `commitment_signed` Message
+
+This message is sent by the opening node. It contains the signatures for
+the first commitment transaction.
+
+Rationale and Requirements are the same as listed above,
+for [`commitment_signed`](#commiting-updates-so-far-commitment_signed) with the following additions.
+
+#### Requirements
+
+The sending node:
+  - MUST send zero HTLC's.
+
+The receiving node:
+  - if the message has one or more HTLC's:
+    - MUST fail the channel.
+  - if it has not already transmittted its `commitment_signed`:
+    - MUST send `commitment_signed`
+  - Otherwise:
+    - MUST send `funding_signed2`
+
+#### Rationale
+
+The first commitment transaction has no HTLC's in it.
+
+Note that the `commitment_signed` message will include the `channel_id` derived from
+the `funding_txid`, instead of the `temporary_node_id`.
+
+
+### The `funding_signed2` Message
+
+This message contains witness data for for the inputs that were
+originally exchanged in `funding_add_input`.
+
+1. type: 64 (`funding_signed2`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`num_witnesses`]
+    * [`num_witnesses*witness_stack`:`witness_stack`]
+
+1. subtype: `witness_stack`
+2. data:
+    * [`sha256`:`prevtx_txid`]
+    * [`u32`:`prevtx_vout`]
+    * [`u16`:`num_input_witness`]
+    * [`num_input_witness*witness_element`:`witness_element`]
+
+1. subtype: `witness_element`
+2. data:
+    * [`u16`:`len`]
+    * [`len*byte`:`witness`]
+
+#### Requirements
+The sending node:
+  - MUST verify it has received valid commitment signatures from its peer
+  - MUST set `witness` to the serialized witness data for the input
+    corresponding to `prevtx_txid`:`prevtx_vout`
+  - MUST remember the details of this funding transaction.
+  - MUST NOT send a `witness_stack` whose length exceeds its
+    previously indicated `max_witness_len`.
+  - if it has NOT received a valid `commitment_signed` message:
+    - MUST NOT send a `funding_signed2` message.
+
+The receiving node:
+  - if the `witness_stack` length exceeds `max_witness_len`:
+    - MUST error.
```

Comment:
What does that mean exactly? Sending an error message to the other party and forgetting everything about this channel creation attempt?

I haven't thought about it deeply enough yet, but I feel that aborting during the funding flow may have some subtle edge cases. Are there some gotchas you ran into while implementing?

Reply:
Hmm. You've highlighted what I believe is the most subjective error case. 

How the opener chooses to handle this is hard to enforce. At this point, they have all the information necessary to successfully broadcast the transaction (the accepter is the only peer that transmits their signatures). If they choose to accept the lower effective feerate anyway and broadcast the tx, there's nothing stopping them. What we're looking to accomplish is incentivize the peer (accepter) to provide an accurate witness length, so that the agreed upon feerate is accurate. The way I've chosen to incentivize this is by threatening that the opener will cancel the open on them.

So yes, the opener would send the peer and error message and, ideally, spend an input to the tx such that the opening transaction was definitively invalidated.